pkg/controller/clustercatalog_controller.go

```diff
@@ -0,0 +1,156 @@
+		return !equality.Semantic.DeepDerivative(required.(*unstructured.Unstructured).UnstructuredContent(), u.UnstructuredContent()), nil
+	}
+}
+
+// applyFunc is a function that is used and expected to perform a
+// server side apply operation. if there is an error during the apply operation,
+// it is returned.
+type applyFunc func(context.Context, types.NamespacedName, string, bool, schema.GroupVersionResource, []byte) error
+
+// managedFunc is a function that returns whether or not the operator
+// is managed. Any errors encountered while evaluating if this operator is
+// managed are returned.
+type managedFunc func() (bool, error)
+
+// shouldUpdateFunc is a function that takes in raw bytes, expected to be a valid
+// Kubernetes resource YAML, and compares it to an *unstructured.Unstructured.
+// if there is a difference between the raw bytes and the *unstructured.Unstructured
+// that warrants an update to the *unstructured.Unstructured, this function will return
+// true with a nil error. Any errors encountered during the evaluation will be returned.
+type shouldUpdateFunc func([]byte, *unstructured.Unstructured) (bool, error)
+
+// getCatalogFunc is a function that gets a ClusterCatalog resource using the provided
+// key. It is returned as a runtime.Object. Any errors encountered during the process
+// of fetching the ClusterCatalog resource will be returned.
+type getCatalogFunc func(types.NamespacedName) (runtime.Object, error)
+
+// clusterCatalogController is a generic controller for managing ClusterCatalog resources.
+//
+// This controller manages resources at the field-level, meaning:
+// - The fields specified in the manifest provided to this controller will always be
+// used. If they are modified by a user on the cluster, they will be reverted by this controller
+// - Any fields not specified in the manifest provided to this controller will not be managed.
+// Users of the cluster are free to modify them as they please.
+type clusterCatalogController struct {
+	name             string
+	key              types.NamespacedName
+	manifest         []byte
+	applyFunc        applyFunc
+	managedFunc      managedFunc
+	shouldUpdateFunc shouldUpdateFunc
+	catalogGetFunc   getCatalogFunc
+}
+
+func (c *clusterCatalogController) sync(ctx context.Context, _ factory.SyncContext) error {
+	logger := klog.FromContext(ctx).WithName(c.name)
+	logger.V(4).Info("sync started")
+	defer logger.V(4).Info("sync finished")
+
+	managed, err := c.managedFunc()
+	if err != nil {
+		return fmt.Errorf("checking if operator is managed: %w", err)
+	}
+	if !managed {
+		return nil
+	}
+
+	catalog, err := c.catalogGetFunc(c.key)
+	if err != nil {
+		return fmt.Errorf("fetching clustercatalog %q: %w", c.key, err)
+	}
+
+	shouldUpdate, err := c.shouldUpdateFunc(c.manifest, catalog.(*unstructured.Unstructured))
+	if err != nil {
+		return fmt.Errorf("determining if clustercatalog %q should be updated: %w", c.key, err)
+	}
+
+	if !shouldUpdate {
+		return nil
```

Comment:
Debug log for when we see that no update is necessary?

Reply:
Should be done in https://github.com/openshift/cluster-olm-operator/pull/73/commits/5b237d4e4d462f19ff7571661b839393bde6b676